File: openpifpaf/decoder/generator/cifcaf.py
from collections import defaultdict
import logging
from queue import PriorityQueue
import time

import numpy as np

from .generator import Generator
from ...annotation import Annotation
from ..field_config import FieldConfig
from ..cif_hr import CifHr
from ..cif_seeds import CifSeeds
from ..caf_scored import CafScored
from .. import nms as nms_module
from ..occupancy import Occupancy
from ... import visualizer

# pylint: disable=import-error
from ...functional import caf_center_s

LOG = logging.getLogger(__name__)


class CifCaf(Generator):
    """Generate CifCaf poses from fields.

    :param: nms: set to None to switch off non-maximum suppression.
    """
    connection_method = 'blend'
    occupancy_visualizer = visualizer.Occupancy()
    force_complete = False
    greedy = False
    keypoint_threshold = 0.0

    def __init__(self, field_config: FieldConfig, *,
                 keypoints,
                 skeleton,
                 out_skeleton=None,
                 confidence_scales=None,
                 worker_pool=None,
                 nms=True):
        super().__init__(worker_pool)
        if nms is True:
            nms = nms_module.Keypoints()

        self.field_config = field_config

        self.keypoints = keypoints
        self.skeleton = skeleton
        self.skeleton_m1 = np.asarray(skeleton) - 1
        self.out_skeleton = out_skeleton or skeleton
        self.confidence_scales = confidence_scales
        self.nms = nms

        self.timers = defaultdict(float)

        # init by_target and by_source
        self.by_target = defaultdict(dict)
        for caf_i, (j1, j2) in enumerate(self.skeleton_m1):
            self.by_target[j2][j1] = (caf_i, True)
            self.by_target[j1][j2] = (caf_i, False)
        self.by_source = defaultdict(dict)
        for caf_i, (j1, j2) in enumerate(self.skeleton_m1):
            self.by_source[j1][j2] = (caf_i, True)
            self.by_source[j2][j1] = (caf_i, False)

    def __call__(self, fields, initial_annotations=None):
        start = time.perf_counter()
        if not initial_annotations:
            initial_annotations = []
        LOG.debug('initial annotations = %d', len(initial_annotations))

        if self.field_config.cif_visualizers:
            for vis, cif_i in zip(self.field_config.cif_visualizers, self.field_config.cif_indices):
                vis.predicted(fields[cif_i])
        if self.field_config.caf_visualizers:
            for vis, caf_i in zip(self.field_config.caf_visualizers, self.field_config.caf_indices):
                vis.predicted(fields[caf_i])

        cifhr = CifHr(self.field_config).fill(fields)
        seeds = CifSeeds(cifhr.accumulated, self.field_config).fill(fields)
        caf_scored = CafScored(cifhr.accumulated, self.field_config, self.skeleton).fill(fields)

        occupied = Occupancy(cifhr.accumulated.shape, 2, min_scale=4)
        annotations = []

        def mark_occupied(ann):
            for joint_i, xyv in enumerate(ann.data):
                if xyv[2] == 0.0:
                    continue

                width = ann.joint_scales[joint_i]
                occupied.set(joint_i, xyv[0], xyv[1], width)  # width = 2 * sigma

        for ann in initial_annotations:
            self._grow(ann, caf_scored)
            annotations.append(ann)
            mark_occupied(ann)

        for v, f, x, y, s in seeds.get():
            if occupied.get(f, x, y):
                continue

            ann = Annotation(self.keypoints, self.out_skeleton).add(f, (x, y, v))
            ann.joint_scales[f] = s
            self._grow(ann, caf_scored)
            annotations.append(ann)
            mark_occupied(ann)

        self.occupancy_visualizer.predicted(occupied)

        LOG.debug('annotations %d, %.3fs', len(annotations), time.perf_counter() - start)

        if self.force_complete:
            annotations = self.complete_annotations(cifhr, fields, annotations)

        if self.nms is not None:
            annotations = self.nms.annotations(annotations)

        LOG.info('%d annotations: %s', len(annotations),
                 [np.sum(ann.data[:, 2] > 0.1) for ann in annotations])
        return annotations

    def _grow_connection(self, xy, xy_scale, caf_field):
        assert len(xy) == 2
        assert caf_field.shape[0] == 9

        # source value
        caf_field = caf_center_s(caf_field, xy[0], xy[1], sigma=2.0 * xy_scale)
        if caf_field.shape[1] == 0:
            return 0, 0, 0, 0

        # source distance
        d = np.linalg.norm(((xy[0],), (xy[1],)) - caf_field[1:3], axis=0)

        # combined value and source distance
        v = caf_field[0]
        sigma = 0.5 * xy_scale
        scores = np.exp(-0.5 * d**2 / sigma**2) * v

        if self.connection_method == 'max':
            return self._target_with_maxscore(caf_field[5:], scores)
        if self.connection_method == 'blend':
            return self._target_with_blend(caf_field[5:], scores)
        raise Exception('connection method not known')

    @staticmethod
    def _target_with_maxscore(target_coordinates, scores):
        assert target_coordinates.shape[1] == scores.shape[0]

        max_i = np.argmax(scores)
        max_entry = target_coordinates[:, max_i]

        score = scores[max_i]
        return max_entry[0], max_entry[1], max_entry[3], score

    @staticmethod
    def _target_with_blend(target_coordinates, scores):
        """Blending the top two candidates with a weighted average.

        Similar to the post processing step in
        "BlazeFace: Sub-millisecond Neural Face Detection on Mobile GPUs".
        """
        assert target_coordinates.shape[1] == len(scores)
        if len(scores) == 1:
            return (
                target_coordinates[0, 0],
                target_coordinates[1, 0],
                target_coordinates[3, 0],
                scores[0] * 0.5,
            )

        sorted_i = np.argsort(scores)
        max_entry_1 = target_coordinates[:, sorted_i[-1]]
        max_entry_2 = target_coordinates[:, sorted_i[-2]]

        score_1 = scores[sorted_i[-1]]
        score_2 = scores[sorted_i[-2]]
        if score_2 < 0.01 or score_2 < 0.5 * score_1:
            return max_entry_1[0], max_entry_1[1], max_entry_1[3], score_1 * 0.5

        # TODO: verify the following three lines have negligible speed impact
        d = np.linalg.norm(max_entry_1[:2] - max_entry_2[:2])
        if d > max_entry_1[3] / 2.0:
            return max_entry_1[0], max_entry_1[1], max_entry_1[3], score_1 * 0.5

        return (
            (score_1 * max_entry_1[0] + score_2 * max_entry_2[0]) / (score_1 + score_2),
            (score_1 * max_entry_1[1] + score_2 * max_entry_2[1]) / (score_1 + score_2),
            (score_1 * max_entry_1[3] + score_2 * max_entry_2[3]) / (score_1 + score_2),
            0.5 * (score_1 + score_2),
        )

    def connection_value(self, ann, caf_scored, start_i, end_i, *, reverse_match=True):
        caf_i, forward = self.by_source[start_i][end_i]
        caf_f, caf_b = caf_scored.directed(caf_i, forward)
        xyv = ann.data[start_i]
        xy_scale_s = max(0.0, ann.joint_scales[start_i])

        new_xysv = self._grow_connection(xyv[:2], xy_scale_s, caf_f)
        keypoint_score = np.sqrt(new_xysv[3] * xyv[2])  # geometric mean
        if keypoint_score < self.keypoint_threshold:
            return 0.0, 0.0, 0.0, 0.0
        if new_xysv[3] == 0.0:
            return 0.0, 0.0, 0.0, 0.0
        xy_scale_t = max(0.0, new_xysv[2])

        # reverse match
        if reverse_match:
            reverse_xyv = self._grow_connection(
                new_xysv[:2], xy_scale_t, caf_b)
            if reverse_xyv[2] == 0.0:
                return 0.0, 0.0, 0.0, 0.0
            if abs(xyv[0] - reverse_xyv[0]) + abs(xyv[1] - reverse_xyv[1]) > xy_scale_s:
                return 0.0, 0.0, 0.0, 0.0

        return (new_xysv[0], new_xysv[1], new_xysv[2], keypoint_score)

    @staticmethod
    def p2p_value(source_xyv, caf_scored, source_s, target_xysv, caf_i, forward):
        caf_f, _ = caf_scored.directed(caf_i, forward)
        xy_scale_s = max(0.0, source_s)

        # source value
        caf_field = caf_center_s(caf_f, source_xyv[0], source_xyv[1],
                                 sigma=2.0 * xy_scale_s)
        if caf_field.shape[1] == 0:
            return 0.0

        # distances
        d_source = np.linalg.norm(
            ((source_xyv[0],), (source_xyv[1],)) - caf_field[1:3], axis=0)
        d_target = np.linalg.norm(
            ((target_xysv[0],), (target_xysv[1],)) - caf_field[5:7], axis=0)

        # combined value and source distance
        xy_scale_t = max(0.0, target_xysv[2])
        sigma_s = 0.5 * xy_scale_s
        sigma_t = 0.5 * xy_scale_t
        scores = (
            np.exp(-0.5 * d_source**2 / sigma_s**2) *
            np.exp(-0.5 * d_target**2 / sigma_t**2) *
            caf_field[0]
        )
        return np.sqrt(source_xyv[2] * max(scores))

    def _grow(self, ann, caf_scored, *, reverse_match=True):
        frontier = PriorityQueue()
        in_frontier = set()

        def add_to_frontier(start_i):
            for end_i, (caf_i, _) in self.by_source[start_i].items():
                if ann.data[end_i, 2] > 0.0:
                    continue
                if (start_i, end_i) in in_frontier:
                    continue

                max_possible_score = np.sqrt(ann.data[start_i, 2])
                if self.confidence_scales is not None:
                    max_possible_score *= self.confidence_scales[caf_i]
                frontier.put((-max_possible_score, None, start_i, end_i))
                in_frontier.add((start_i, end_i))
                ann.frontier_order.append((start_i, end_i))

        def frontier_get():
            while frontier.qsize():
                entry = frontier.get()
                if entry[1] is not None:
                    return entry

                _, __, start_i, end_i = entry
                if ann.data[end_i, 2] > 0.0:
                    continue

                new_xysv = self.connection_value(
                    ann, caf_scored, start_i, end_i, reverse_match=reverse_match)
                if new_xysv[3] == 0.0:
                    continue
                score = new_xysv[3]
                if self.greedy:
                    return (-score, new_xysv, start_i, end_i)
                if self.confidence_scales is not None:
                    caf_i, _ = self.by_source[start_i][end_i]
                    score *= self.confidence_scales[caf_i]
                frontier.put((-score, new_xysv, start_i, end_i))

        # seeding the frontier
        for joint_i, v in enumerate(ann.data[:, 2]):
            if v == 0.0:
                continue
            add_to_frontier(joint_i)

        while True:
            entry = frontier_get()
            if entry is None:
                break

            _, new_xysv, jsi, jti = entry
            if ann.data[jti, 2] > 0.0:
                continue

            ann.data[jti, :2] = new_xysv[:2]
            ann.data[jti, 2] = new_xysv[3]
            ann.joint_scales[jti] = new_xysv[2]
            ann.decoding_order.append(
                (jsi, jti, np.copy(ann.data[jsi]), np.copy(ann.data[jti])))
            add_to_frontier(jti)

    def _flood_fill(self, ann):
        frontier = PriorityQueue()

        def add_to_frontier(start_i):
            for end_i in self.by_source[start_i].keys():
                if ann.data[end_i, 2] > 0.0:
                    continue
                start_xyv = ann.data[start_i].tolist()
                frontier.put((-xyv[2], end_i, start_xyv, ann.joint_scales[start_i]))

        for start_i, xyv in enumerate(ann.data):
            if xyv[2] == 0.0:
                continue
            add_to_frontier(start_i)

        while frontier.qsize():
            _, end_i, xyv, s = frontier.get()
            if ann.data[end_i, 2] > 0.0:
                continue
            ann.data[end_i, :2] = xyv[:2]
            ann.data[end_i, 2] = 0.00001
            ann.joint_scales[end_i] = s
            add_to_frontier(end_i)

    def complete_annotations(self, cifhr, fields, annotations):
        start = time.perf_counter()

        caf_scored = CafScored(cifhr.accumulated, self.field_config, self.skeleton,
                               score_th=0.0001).fill(fields)

        for ann in annotations:
            unfilled_mask = ann.data[:, 2] == 0.0
            self._grow(ann, caf_scored, reverse_match=False)
            now_filled_mask = ann.data[:, 2] > 0.0
            updated = np.logical_and(unfilled_mask, now_filled_mask)
            ann.data[updated, 2] = np.minimum(0.001, ann.data[updated, 2])

            # some joints might still be unfilled
            if np.any(ann.data[:, 2] == 0.0):
                self._flood_fill(ann)

        LOG.debug('complete annotations %.3fs', time.perf_counter() - start)
        return annotations

File: cpp/examples/cifcafdecoder_v0.12.13_wholebody/src/decoder/cif_caf.cpp
#include <cmath>

#include "../../include/decoder/caf_scored.hpp"
#include "../../include/decoder/functional.hpp"
#include "../../include/decoder/annotation.hpp"
#include "../../include/decoder/cif_seeds.hpp"
#include "../../include/decoder/occupancy.hpp"
#include "../../include/decoder/cif_caf.hpp"
#include "../../include/decoder/cif_hr.hpp"
#include "../../include/utils/vector_utils.hpp"

struct PreprocessingMeta {
    std::vector<float> offset;
    std::vector<float> scale;
    std::vector<float> valid_area;
    std::vector<int> width_height;

    bool hflip{false};
    float rotation_angle{0};
    float rotation_width{-1};
    float rotation_height{-1};
};

CifCaf::CifCaf() {}

CifCaf::CifCaf(
    float *field1,
    float *field2,
    unsigned H,
    unsigned W,
    const vector<float> &confidence_scales,
    bool nms
) {
    this->fields = Fields {field1, field2};
    this->H = H;
    this->W = W;
    this->nms = nms;

    vector<pair<int, int>> sk = this->cif_metas.head_metas.skeleton;
    for(int i = 0; i < sk.size(); i++) {
        sk[i].first -= 1;
        sk[i].second -= 1;
    }
    this->skeleton_m1 = sk;
    this->keypoints = this->cif_metas.head_metas.keypoints;
    this->score_weights = this->cif_metas.head_metas.score_weights;
    this->out_skeleton = this->caf_metas.head_metas.skeleton;
    this->confidence_scales = confidence_scales;

    this->priority += this->cif_metas.n_fields / 1000.0;
    this->priority += this->caf_metas.n_fields / 1000.0;

    for(int caf_i = 0; caf_i < this->skeleton_m1.size(); caf_i++) {
        int j1 = this->skeleton_m1[caf_i].first;
        int j2 = this->skeleton_m1[caf_i].second;
        this->by_target[j2][j1] = make_pair(caf_i, true);
        this->by_target[j1][j2] = make_pair(caf_i, false);
    }
    for(int caf_i = 0; caf_i < this->skeleton_m1.size(); caf_i++) {
        int j1 = this->skeleton_m1[caf_i].first;
        int j2 = this->skeleton_m1[caf_i].second;
        this->by_source[j1][j2] = make_pair(caf_i, true);
        this->by_source[j2][j1] = make_pair(caf_i, false);
    }
}

vector<Annotation> CifCaf::decode(vector<Annotation> &initial_annotations) {
    if(!initial_annotations.empty())
        vector<Annotation> initial_annotations;

    CifHr cif_hr(this->H, this->W, this->cif_metas);
    cif_hr.fill(this->fields);
    CifSeeds cif_seeds(
        cif_hr.accumulated,
        this->H, this->W
    );
    cif_seeds.fill(this->fields);

    CafScored caf_scored = CafScored(
        cif_hr.accumulated, this->H, this->W, this->caf_metas
    ).fill(this->fields);

    int d1 = cif_hr.accumulated.shape[0];
    int d2 = cif_hr.accumulated.shape[1];
    int d3 = cif_hr.accumulated.shape[2];
    vector<int> cif_hr_accumulated_shape {d1, d2, d3};
    Occupancy occupied = Occupancy(cif_hr_accumulated_shape, 2.0, 4.0);

    vector<Annotation> annotations;
    for(int i = 0; i < initial_annotations.size(); i++) {
        this->_grow(initial_annotations[i], caf_scored);
        annotations.push_back(initial_annotations[i]);
        this->mark_occupied(initial_annotations[i], occupied);
    }

    vector<Seed> seeds = cif_seeds.get();
    // print_vector(seeds);
    for(int i = 0; i < seeds.size(); i++) {
        float v = seeds[i].vv;
        int f = seeds[i].field;
        float x = seeds[i].xx;
        float y = seeds[i].yy;
        float s = seeds[i].ss;

        if(occupied.get(f, x, y)) {
            continue;
        }

        vector<float> xyv {x, y, v};
        Annotation ann = Annotation(this->keypoints, this->out_skeleton, this->score_weights).add(f, xyv);
        ann.joint_scales[f] = s;
        this->_grow(ann, caf_scored);
        annotations.push_back(ann);
        this->mark_occupied(ann, occupied);
    }

    if(this->force_complete) {
        // TODO implement
    }

    vector<Annotation> final_annotations;
    if(this->nms) {
        final_annotations = this->nms_keypoints.annotations(annotations);
    }

    return final_annotations;
}

void CifCaf::mark_occupied(Annotation &ann, Occupancy &occupied) {
    vector<int> joint_is;

    int rows = ann.data.size();
    for(int i = 0; i < rows; i ++) {
        if(ann.data[i][2] != 0.0)
            joint_is.push_back(i);
    }

    for(int i = 0; i < joint_is.size(); i++) {
        int joint_i = joint_is[i];
        float width = ann.joint_scales[i];

        occupied.set(joint_i, ann.data[joint_i][0], ann.data[joint_i][1], width);
    }
}

void CifCaf::_grow(Annotation &ann, CafScored &caf_scored, bool reverse_match) {
    priority_queue<HeapItem, vector<HeapItem>, heap_item_comparer> frontier;
    set<pair<int, int>> in_frontier;

    int rows = ann.data.size();
    vector<int> joint_is;
    for(int i = 0; i < rows; i ++) {
        if(ann.data[i][2] != 0.0)
            joint_is.push_back(i);
    }
    for(int joint_i = 0; joint_i < joint_is.size(); joint_i++) {
        this->_add_to_frontier(frontier, in_frontier, ann, joint_is[joint_i]);
    }

    HeapItem entry;
    while(true) {
        entry = this->_frontier_get(frontier, ann, caf_scored, reverse_match);
        if(entry.new_xysv.empty())
            break;

        vector<float> new_xysv = entry.new_xysv;
        int jsi = entry.start_i;
        int jti = entry.end_i;
        if(ann.data[jti][2] > 0.0)
            continue;

        ann.data[jti][0] = new_xysv[0];
        ann.data[jti][1] = new_xysv[1];
        ann.data[jti][2] = new_xysv[3];
        ann.joint_scales[jti] = new_xysv[2];
        struct DecodingStep decst {
            jsi, jti, ann.data[jsi], ann.data[jti]
        };
        ann.decoding_order.push_back(decst);

        this->_add_to_frontier(frontier, in_frontier, ann, jti);
    }
}


void CifCaf::_add_to_frontier(
    priority_queue<HeapItem, vector<HeapItem>, heap_item_comparer> &frontier,
    set<pair<int, int>> &in_frontier,
    Annotation &ann,
    int start_i
) {
    map<int, pair<int, bool>> by_source_items = this->by_source[start_i];

    map<int, pair<int, bool>>::iterator it;
    for(it = by_source_items.begin(); it != by_source_items.end(); it++) {
        int end_i = it->first;
        pair<int, bool> pair_ = it->second;
        int caf_i = pair_.first;

        if(ann.data[end_i][2] > 0.0)
            continue;

        if(in_frontier.count(make_pair(start_i, end_i)))
            continue;

        float max_possible_score = ann.data[start_i][2];
        if(!this->confidence_scales.empty())
            max_possible_score *= this->confidence_scales[caf_i];

        vector<float> empty;
        struct HeapItem heap_vals {-max_possible_score, empty, start_i, end_i};
        frontier.push(heap_vals);
        in_frontier.insert(make_pair(start_i, end_i));
        ann.frontier_order.push_back(make_pair(start_i, end_i));
    }
}

HeapItem CifCaf::_frontier_get(
    priority_queue<HeapItem, vector<HeapItem>, heap_item_comparer> &frontier,
    Annotation &ann,
    CafScored &caf_scored,
    bool reverse_match
) {
    if(frontier.empty()) {
        return HeapItem {-1.0, vector<float>(), -1, -1};
    }

    while(!frontier.empty()) {
        struct HeapItem entry = frontier.top();

        frontier.pop();

        if(!entry.new_xysv.empty()) {
           return entry;
        }

        int start_i = entry.start_i;
        int end_i = entry.end_i;
        if(ann.data[end_i][2] > 0.0) {
            continue;
        }

        vector<float> new_xysv = this->connection_value(ann, caf_scored, start_i, end_i, reverse_match);

        if(new_xysv[3] == 0.0)
            continue;

        float score = new_xysv[3];
        if(this->greedy)
            return HeapItem {-score, new_xysv, start_i, end_i};

        if(!this->confidence_scales.empty()) {
            pair<int, bool> pair_ = this->by_source[start_i][end_i];
            int caf_i = pair_.first;
            score *= this->confidence_scales[caf_i];
        }
        HeapItem v {-score, new_xysv, start_i, end_i};
        frontier.push(v);
    }

    if(frontier.empty()) {
        return HeapItem {-1.0, vector<float>(), -1, -1};
    }

    //TODO return something in default case to silent compiler warning
}

vector<float> CifCaf::connection_value(
    Annotation &ann,
    CafScored &caf_scored,
    int start_i,
    int end_i,
    bool reverse_match
) {
    pair<int, bool> pair_ = this->by_source[start_i][end_i];
    int caf_i = pair_.first;
    bool forward = pair_.second;

    pair<Vector2D, Vector2D> directed = caf_scored.directed(caf_i, forward);
    Vector2D caf_f = directed.first;
    Vector2D caf_b = directed.second;

    vector<float> xyv = ann.data[start_i];
    float xy_scale_s = max(0.0f, ann.joint_scales[start_i]);

    bool only_max;
    if(this->connection_method.compare("max") == 0)
        only_max = true;
    else
        only_max = false;

    vector<float> new_xysv = grow_connection_blend(caf_f, xyv[0], xyv[1], xy_scale_s, only_max);
    if(new_xysv[3] == 0.0)
        return vector<float> {0.0, 0.0, 0.0, 0.0};

    float keypoint_score = sqrt(new_xysv[3] * xyv[2]);

    if(keypoint_score < xyv[2] * this->keypoint_threshold_rel)
        return vector<float> {0.0, 0.0, 0.0, 0.0};

    float xy_scale_t = max(0.0f, new_xysv[2]);
    if(reverse_match) {
        vector<float> reverse_xyv = grow_connection_blend(caf_b, new_xysv[0], new_xysv[1], xy_scale_t, only_max);
        if(reverse_xyv[2] == 0.0)
            return vector<float> {0.0, 0.0, 0.0, 0.0};
        if(abs(xyv[0] - reverse_xyv[0]) + abs(xyv[1] - reverse_xyv[1]) > xy_scale_s)
            return vector<float> {0.0, 0.0, 0.0, 0.0};
    }

    return vector<float> {new_xysv[0], new_xysv[1], new_xysv[2], keypoint_score};
}


void CifCaf::annotation_inverse(vector<Annotation> &annotations, const struct PreprocessingMeta &meta) {
    float angle = -meta.rotation_angle;
    float rw = meta.rotation_width;
    float rh = meta.rotation_height;

    float cangle = cos(angle / 180.0 + M_PI);
    float sangle = sin(angle / 180.0 + M_PI);

    for(int i = 0; i < annotations.size(); i++) {
        if(angle != 0.0) {
            for(int j = 0; j < annotations[i].data.size(); j++) {
                float x_old = annotations[i].data[j][0] - (rw - 1) / 2;
                float y_old = annotations[i].data[j][1] - (rh - 1) / 2;

                annotations[i].data[j][0] = (rw - 1) / 2 + cangle * x_old + sangle * y_old;
                annotations[i].data[j][1] = (rh - 1) / 2 - sangle * x_old + cangle * y_old;
            }
        }

        for(int j = 0; j < annotations[i].data.size(); j++) {
            annotations[i].data[j][0] += meta.offset[0];
            annotations[i].data[j][1] += meta.offset[1];
        }

        for(int j = 0; j < annotations[i].data.size(); j++) {
            annotations[i].data[j][0] /= meta.scale[0];
            annotations[i].data[j][1] /= meta.scale[1];
        }

        for(int j = 0; j < annotations[i].joint_scales.size(); j++) {
            annotations[i].joint_scales[j] /= meta.scale[0];
        }

        if(meta.hflip) {
            int w = meta.width_height[0];
            for(int j = 0; j < annotations[i].data.size(); j++) {
                annotations[i].data[j][0] = -annotations[i].data[j][0] + (w - 1);
            }
        }

        for(int j = 0; j < annotations[i].decoding_order.size(); j++) {
            annotations[i].decoding_order[j].jsi_vec[0] += meta.offset[0];
            annotations[i].decoding_order[j].jsi_vec[1] += meta.offset[1];

            annotations[i].decoding_order[j].jti_vec[0] += meta.offset[0];
            annotations[i].decoding_order[j].jti_vec[1] += meta.offset[1];

            annotations[i].decoding_order[j].jsi_vec[0] /= meta.scale[0];
            annotations[i].decoding_order[j].jsi_vec[1] /= meta.scale[1];

            annotations[i].decoding_order[j].jti_vec[0] /= meta.scale[0];
            annotations[i].decoding_order[j].jti_vec[1] /= meta.scale[1];
        }
    }
}
